sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -336,9 +383,15 @@ object DecorrelateInnerQuery extends PredicateHelper {
         throw new IllegalStateException(
           s"Unable to rewrite domain join with conditions: $conditions\n$d.")
       }
-    case u: Union =>
-      u.mapChildren { child =>
-        rewriteDomainJoins(outerPlan, child, pushConditionsThroughUnion(conditions, u, child))
+    case s @ (_ : Union | _: SetOperation) =>
+      s.mapChildren { child =>
+        rewriteDomainJoins(outerPlan, child,
+          pushDomainConditionsThroughSetOperation(conditions, s, child))
```

Comment:
It would be great to provide a simple example here to illustrate the change for `conditions` (before vs after pushDomainConditionsThroughSetOperation)